### ISSUE TITLE
About Reproduction

### ISSUE THREAD

**Issue by g1moon** [locuslab/open-unlearning#71]
### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task
- [ ] My own task or dataset (give details below)

### Reproduction

Hello! First, thank you very much for sharing your excellent research and code.

After installing the environment exactly as provided in the repository, I ran the tofu_unlearning.sh script on Llama-3.2-1B-Instruct without changing any settings. The results I obtained were as follows:

```
=== Llama-3.2-1B-Instruct Results for forget_01 ===
Method       forget_quality model_utility  forget_truth_ratio
finetuned    0.01430        0.59832        0.47464       
retain       1.00           0.59582        0.64756       
GradAscent   0.00676        0.59571        0.48166       
GradDiff     0.00676        0.60041        0.47346       
NPO          0.00676        0.59667        0.48255       
DPO          0.00302        0.59121        0.48997       
RMU          0.26569        0.46797        0.57000       

=== Llama-3.2-1B-Instruct Results for forget_05 ===
Method       forget_quality model_utility  forget_truth_ratio
finetuned    2.96e-13       0.59832        0.47333       
retain       1.00           0.59843        0.63361       
GradAscent   0.01184        0.00000        0.66449       
GradDiff     0.00028        0.30374        0.47527       
NPO          0.00301        0.42221        0.63833       
DPO          2.41e-08       0.03565        0.58899       
RMU          0.00609        0.50589        0.60112       

=== Llama-3.2-1B-Instruct Results for forget_10 ===
Method       forget_quality model_utility  forget_truth_ratio
finetuned    8.08e-22       0.59832        0.47619       
retain       1.00           0.59326        0.62804       
GradAscent   2.73e-12       0.00000        0.78945       
GradDiff     6.39e-06       0.00000        0.75138       
NPO          4.58e-07       0.25581        0.68759       
DPO          3.28e-14       0.23112        0.59212       
RMU          1.75e-09       0.48745        0.57433 

```

Comparing to the results in results.md, I was able to reproduce the finetuned and Retain values exactly, but the rest of the baselines yielded suspicious numbers. 

If the setup described in results.md is indeed the same as the default settings in your GitHub repository, then the only differences I can see are:
	1.	I used two A100 GPUs,
	2.	I set per_device_train_batch_size=32.
However, I also tried per_device_train_batch_size=4 and got the same results.

Are there any specific settings or details I should check for reproducing the results? Or is it possible that the difference in hardware (GPU) might be causing this discrepancy? I’d greatly appreciate any advice you can offer.

### Expected behavior

https://github.com/locuslab/open-unlearning/blob/main/docs/results.md

**Comment by molereddy** [comment]
Hi, we find that this setup is vulnerable to changes in training configurations. Several unlearning methods work better only at tuned parameters. For example, we found that sometimes we had to 5x or 2x the learning rate reported in the original TOFU, NPO papers resp to get similar results. We also found sensitivity to deepspeed usage, batch size etc.
1. Can you elaborate and share all your training parameters?
2. You may try tuning the hyperparameters for each method to get better results.

**Comment by molereddy** [comment]
It would be great if you could dump the config file which would've been created in your experiment's directory. And do elaborate on your hardware setup: are you using 40G or 80G A100s?

**Comment by then-go** [comment]
Thank you for your responding.

1. About training parameter.
I used the default training parameters provided in the GitHub repository(e.g. unlearn/tofu/default.yaml.). The contents of unlearn/tofu/defaul.yaml are shown below:

```
# @package _global_

defaults:
  - override /model: Llama-3.2-1B-Instruct
  - override /trainer: GradAscent
  - override /data: unlearn
  - override /data/datasets@data.forget: TOFU_QA_forget
  - override /data/datasets@data.retain: TOFU_QA_retain
  - override /eval: tofu

model:
  model_args:
    pretrained_model_name_or_path: open-unlearning/tofu_Llama-3.2-1B-Instruct_full

forget_split: forget10
retain_split: retain90
retain_logs_path: null

eval:
  tofu:
    forget_split: ${forget_split}
    retain_logs_path: ${retain_logs_path}
    overwrite: true
    
data:
  anchor: forget
  forget:
    TOFU_QA_forget: 
      args:
        hf_args:
          name: ${forget_split}
  retain:
    TOFU_QA_retain:
      args:
        hf_args:
          name: ${retain_split}

trainer:
  args:
    warmup_epochs: 1.0 # custom parameter
    learning_rate: 1e-5
    weight_decay: 0.01
    num_train_epochs: 10
    # save_strategy: steps
    # save_steps: 0.5

task_name: ???
```


2. About my hardware.
I used two A100 80g hardware.

I’m aware that unlearning methods can be sensitive to hyperparameters, but I don’t believe their performance should hinge on hardware differences. Do you think the discrepancy in reproducing results might be caused by the difference in hardware?

**Comment by molereddy** [comment]
We are investigating this discrepancy.
Can you share the final filled up config for any one of your experiments that is mismatched? We want to make sure all things match exactly. What you've shared is the pre-override experiment input config.
When your experiment runs, it creates a directory which contains a `.hydra/config.yaml` which contains the entire config. Can you find and share that?

**Comment by then-go** [comment]
Thank you very much for your assistance. Below is my config.yml file for forget01.

[config_forget01.zip](https://github.com/user-attachments/files/19223572/config_forget01.zip)

**Comment by Dornavineeth** [comment]
Thank you for sharing the config files. The discrepancy in results appears to be due to the batch size. When `per_device_train_batch_size` is set to `32`, I obtain results very similar to yours.

You mentioned that the results remain unchanged even when `per_device_train_batch_size=4`. This is likely because the results are **not being overwritten** during evaluation, as `overwrite=false` is set by default in the eval config, though the model checkpoint is different now. Could you please delete the `evals` folder for each experiment and re-run the evaluation after training the model with `per_device_train_batch_size=4`.

This should yield improved results for the `forget01` split. However, some variations may still persist compared to the values reported in `results.md` due to differences in hardware and system configurations.

**Note:** Even with a fixed random seed and same hardware, results may not be **exactly** identical across machines.
Have a look at this [discussion](https://discuss.huggingface.co/t/getting-different-results-on-different-hardware/720).

Hope this helps!

**Comment by then-go** [comment]
oh, I missed the fact that per_device_train_batch_size=4 result is not overwrited.

I ran the experiment again with per_device_train_batch_size=4, and here are the outcomes:
```
=== Llama-3.2-1B-Instruct Results for forget_01 ===
Method       forget_quality model_utility  forget_truth_ratio
finetuned    0.01           0.60           0.47          
retain       1.00           0.60           0.65          
GradAscent   0.10           0.46           0.64          
GradDiff     0.77           0.41           0.59          
NPO          0.77           0.53           0.66          
DPO          0.01           0.50           0.60          
RMU          0.05           0.55           0.69    

```

The results differ slightly from those in your result.md, but I believe such a minor discrepancy is reasonable. 
Thank you again for all your help!

**Comment by NilakshanKunananthaseelan** [comment]
Thanks for the excellent work. OpenUnlearning made it easy to benchmark results. I'm trying to add my custom method, however I have a similar discrepancy for `model utility` and `forget quality` when using forget01 split, while the` forget_truth_ration` has closer to the reported value for llama-3.2-1B-Instruct.

My hardware details: **NVIDIA RTX A5500. Num GPUs = 2. Memory: 24GB. Platform: Linux.**
Here is my script:
```
#!/bin/bash


export MASTER_PORT=$(python -c "import socket; s=socket.socket(); s.bind(('', 0)); print(s.getsockname()[1]); s.close()")
echo "Master Port: $MASTER_PORT"

models=(
    "Llama-3.2-1B-Instruct"
    # "Llama-3.2-3B-Instruct"
    # "Llama-3.1-8B-Instruct"
)
trainers_experiments=(
    "GradAscent unlearn/tofu/default.yaml"
    "GradDiff unlearn/tofu/default.yaml"
    "NPO unlearn/tofu/default.yaml"
    "DPO unlearn/tofu/idk.yaml"
    "RMU  unlearn/tofu/default.yaml"
)
splits=(
    "forget01 holdout01 retain99"
    "forget05 holdout05 retain95"
    "forget10 holdout10 retain90"
)


per_device_train_batch_size=4 # on two gpus would make effective batch size 32
gradient_accumulation_steps=4


########################################################################################################################
########################################### Unlearn TOFU models ########################################################
########################################################################################################################


for split in "${splits[@]}"; do
    forget_split=$(echo $split | cut -d' ' -f1)
    holdout_split=$(echo $split | cut -d' ' -f2)
    retain_split=$(echo $split | cut -d' ' -f3)

    for model in "${models[@]}"; do
        for trainer_experiment in "${trainers_experiments[@]}"; do
            trainer=$(echo $trainer_experiment | cut -d' ' -f1)
            experiment=$(echo $trainer_experiment | cut -d' ' -f2)
            echo "--------------------------------"
            echo "trainer: $trainer"
            echo "experiment: $experiment"
            echo "forget_split: $forget_split"
            echo "holdout_split: $holdout_split"
            echo "retain_split: $retain_split"
            echo "--------------------------------"
            
            task_name=tofu_${model}_${forget_split}_${trainer} 
            model_path=/home/nilakshan/0-Unlearning/0-open-unlearning-dev/saves/finetune/tofu_${model}_full
            echo ${task_name}: Unlearning ${model_path} using ${trainer}

            # Unlearn
            CUDA_VISIBLE_DEVICES=0,1 accelerate launch --config_file configs/accelerate/default_config.yaml --main_process_port $MASTER_PORT \
            src/train.py --config-name=unlearn.yaml \
            experiment=${experiment} \
            trainer=${trainer} \
            task_name=${task_name} \
            model=${model} \
            forget_split=${forget_split} \
            retain_split=${retain_split} \
            model.model_args.pretrained_model_name_or_path=${model_path} \
            retain_logs_path=saves/eval/tofu_${model}_${retain_split}/TOFU_EVAL.json \
            trainer.args.per_device_train_batch_size=$per_device_train_batch_size \
            trainer.args.gradient_accumulation_steps=$gradient_accumulation_steps \
            trainer.args.ddp_find_unused_parameters=true \
            trainer.args.gradient_checkpointing=true

            # Eval
            CUDA_VISIBLE_DEVICES=0 python src/eval.py \
            experiment=eval/tofu/default.yaml \
            forget_split=${forget_split} \
            holdout_split=${holdout_split} \
            model=${model} \
            task_name=${task_name} \
            model.model_args.pretrained_model_name_or_path=saves/unlearn/${task_name} \
            paths.output_dir=saves/unlearn/${task_name}/evals \
            retain_logs_path=saves/eval/tofu_${model}_${retain_split}/TOFU_EVAL.json
        done
    done
done
```

Here is the tofu unlearn config:
```

# @package _global_

defaults:
  - override /model: Llama-3.2-1B-Instruct
  - override /trainer: GradAscent
  - override /data: unlearn
  - override /data/datasets@data.forget: TOFU_QA_forget
  - override /data/datasets@data.retain: TOFU_QA_retain
  - override /eval: tofu

model:
  model_args:
    pretrained_model_name_or_path: /home/nilakshan/0-Unlearning/0-open-unlearning-dev/saves/finetune/tofu_Llama-3.2-1B-Instruct_full
forget_split: forget10
retain_split: retain90
holdout_split: holdout10
retain_logs_path: null
question_key: "question"

eval:
  tofu:
    forget_split: ${forget_split}
    holdout_split: ${holdout_split}
    retain_logs_path: ${retain_logs_path}
    overwrite: true
    question_key: ${question_key}
    
data:
  anchor: forget
  forget:
    TOFU_QA_forget: 
      args:
        hf_args:
          name: ${forget_split}
  retain:
    TOFU_QA_retain:
      args:
        hf_args:
          name: ${retain_split}

trainer:
  args:
    warmup_epochs: 1.0 # custom parameter
    learning_rate: 1e-5
    weight_decay: 0.01
    num_train_epochs: 10
    # save_strategy: steps
    # save_steps: 0.5

task_name: ???
```

```
Package                  Version     Build
------------------------ ----------- -----
absl-py                  2.3.1
accelerate               0.34.2
aiohappyeyeballs         2.6.1
aiohttp                  3.12.15
aiosignal                1.4.0
annotated-types          0.7.0
antlr4-python3-runtime   4.9.3
attrs                    25.3.0
bitsandbytes             0.44.1
certifi                  2025.8.3
chardet                  5.2.0
charset-normalizer       3.4.3
click                    8.2.1
colorama                 0.4.6
colorlog                 6.9.0
DataProperty             1.1.0
datasets                 3.0.1
deepspeed                0.15.4
dill                     0.3.8
evaluate                 0.4.5
filelock                 3.18.0
frozenlist               1.7.0
fsspec                   2024.6.1
geoopt                   0.5.1
grpcio                   1.74.0
hjson                    3.1.0
huggingface-hub          0.29.1
hydra-colorlog           1.2.0
hydra-core               1.3.0
idna                     3.10
Jinja2                   3.1.6
joblib                   1.5.1
jsonlines                4.0.0
lm_eval                  0.4.8
lxml                     6.0.0
Markdown                 3.8.2
MarkupSafe               3.0.2
mbstrdecoder             1.1.4
more-itertools           10.7.0
mpmath                   1.3.0
msgpack                  1.1.1
multidict                6.6.4
multiprocess             0.70.16
networkx                 3.5
ninja                    1.13.0
nltk                     3.9.1
numexpr                  2.11.0
numpy                    2.2.3
nvidia-cublas-cu12       12.1.3.1
nvidia-cuda-cupti-cu12   12.1.105
nvidia-cuda-nvrtc-cu12   12.1.105
nvidia-cuda-runtime-cu12 12.1.105
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu12        11.0.2.54
nvidia-curand-cu12       10.3.2.106
nvidia-cusolver-cu12     11.4.5.107
nvidia-cusparse-cu12     12.1.0.106
nvidia-nccl-cu12         2.20.5
nvidia-nvjitlink-cu12    12.9.86
nvidia-nvtx-cu12         12.1.105
omegaconf                2.3.0
open-unlearning          0.1.0
packaging                25.0
pandas                   2.3.1
pathvalidate             3.3.1
peft                     0.17.0
pip                      25.1
portalocker              3.2.0
POT                      0.9.5
propcache                0.3.2
protobuf                 6.31.1
psutil                   7.0.0
py-cpuinfo               9.0.0
pyarrow                  21.0.0
pybind11                 3.0.0
pydantic                 2.11.7
pydantic_core            2.33.2
pytablewriter            1.2.1
python-dateutil          2.9.0.post0
pytz                     2025.2
PyYAML                   6.0.2
regex                    2025.7.34
requests                 2.32.4
rouge_score              0.1.2
sacrebleu                2.5.1
safetensors              0.6.2
scikit-learn             1.5.2
scipy                    1.14.1
setuptools               78.1.1
six                      1.17.0
sqlitedict               2.1.0
sympy                    1.14.0
tabledata                1.3.4
tabulate                 0.9.0
tcolorpy                 0.1.7
tensorboard              2.18.0
tensorboard-data-server  0.7.2
threadpoolctl            3.6.0
tokenizers               0.20.3
torch                    2.4.1
tqdm                     4.67.1
tqdm-multiprocess        0.0.11
transformers             4.45.1
triton                   3.0.0       1
typepy                   1.3.4
typing_extensions        4.14.1
typing-inspection        0.4.1
tzdata                   2025.2
urllib3                  2.5.0
Werkzeug                 3.1.3
wheel                    0.45.1
word2number              1.1
xxhash                   3.5.0
yarl                     1.20.1
zstandard                0.23.0
```

Here are the results for forget01 split:
```
Forget01

forget_quality  model_utility
GradAscent        0.918805       0.454957
GradDiff          0.265687       0.465242
RMU               0.097075       0.554990
NPO               0.765931       0.563150


While other splits have closer results:

Forget05

forget_quality  model_utility
GradAscent   1.942643e-119       0.000000
GradDiff     1.942643e-119       0.507240
RMU           1.385084e-11       0.584067
NPO           1.420747e-01       0.447978

Forget10
forget_quality  model_utility
GradAscent   1.063590e-239       0.000000
GradDiff     9.041959e-232       0.471667
RMU           3.277026e-14       0.585462
NPO           1.959480e-02       0.448172

```

And, for the DPO where should I download the  `./data/idk.jsonl` file.
